src/scarlet2/validation.py

```diff
@@ -1,46 +1,217 @@
     """
-    if observation is None:
-        return
 
-    pass
+    validation_errors = []
+
+    observation_checks = ObservationCheck.get_checks()
+    for check in observation_checks:
+        if error := check(observation).check():
+            validation_errors.append(error)
+
+    return validation_errors
 
 
-def check_scene(scene: Optional[Scene] = None):
+def check_scene(scene: Scene) -> list[ValidationError]:
     """Check the scene against the various validation rules.
 
     Parameters
     ----------
-    scene : Scene, optional
-        The scene object to check, by default None
+    scene : Scene
+        The scene object to check.
+
+    Returns
+    -------
+    list[ValidationError]
+        A list of validation errors found in the source objects in the scene.
+        If no errors are found, the list is empty.
+    """
+
+    validation_errors = []
+    for source in scene.sources:
+        validation_errors.extend(check_source(source))
+
+    return validation_errors
+
+
+def check_source(source: Source) -> list[ValidationError]:
+    """Check the source against the various validation rules.
+
+    Parameters
+    ----------
+    source : Source
+        The source object to check.
+
+    Returns
+    -------
+    list[ValidationError]
+        A list of validation errors found in the source object.
+        If no errors are found, the list is empty.
     """
-    if scene is None:
-        return
 
-    pass
+    validation_errors = []
+
+    # Perform various checks on the source object
+    source_checks = SourceCheck.get_checks()
+    for check in source_checks:
+        if error := check(source).check():
+            validation_errors.append(error)
+
+    return validation_errors
 
 
-def check_fit(scene: Optional[Scene] = None):
+def check_fit(scene: Scene) -> list[ValidationError]:
     """Check the scene after fitting against the various validation rules.
 
     Parameters
     ----------
-    scene : Scene, optional
-        The scene object to check, by default None
+    scene : Scene
+        The scene object to check.
+
+    Returns
+    -------
+    list[ValidationError]
+        A list of validation errors found in the fit of the scene.
+        If no errors are found, the list is empty.
     """
-    if scene is None:
-        return
 
-    pass
+    validation_errors = []
+    fit_checks = FitCheck.get_checks()
+    for check in fit_checks:
+        validation_errors.append(check(scene).check())
+
+    return validation_errors
+
+
+class BaseCheck:
+    """A base class for checks."""
+
+    def __init__(self):
+        pass
+
+    @classmethod
+    def get_checks(cls):
+        """Get the list of all registered checks."""
+        if not hasattr(cls, "_checks"):
+            cls._checks = []
+        return cls._checks
+
+    def check(self) -> Optional[ValidationError]:
+        """Run the check on the object."""
+        raise NotImplementedError("Subclasses should implement this method.")
```

Comment:
That feels somewhat heavy to me, and the sense that we need to create a class for each object we seek to test. I _thinkk_ we could do this simply with pytest-like functions.